File: Problem-Solving/LeetCode/0133-clone-graph/0133-clone-graph.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Neighbors []*Node
 * }
 */

func cloneGraph(node *Node) *Node {
    if node == nil {
        return nil
    }
    copiedNodes := make([]*Node, 101)
    return dfs(node, copiedNodes)
}

func dfs(node *Node, copiedNodes []*Node) *Node {
    if copiedNodes[node.Val] != nil {
        return copiedNodes[node.Val]
    }
    copiedNode := &Node {
        Val: node.Val,
    }
    copiedNodes[node.Val]= copiedNode
    for _, neighbor := range node.Neighbors {
        copiedNeighbor := dfs(neighbor, copiedNodes)
        copiedNode.Neighbors = append(copiedNode.Neighbors, copiedNeighbor)
    }
    return copiedNode
}

File: Problem-Solving/LeetCode/0133-clone-graph/0133-clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    vector<Node*> clonedNodes;
    Node* cloneGraph(Node* node) {
        clonedNodes.assign(101, nullptr);
        return cloneDfs(node, clonedNodes);
    }

    Node* cloneDfs(Node* node, vector<Node*>& clonedNodes) {
        if (node == nullptr) return nullptr;
        if (clonedNodes[node->val]) {
            return clonedNodes[node->val];
        }
        Node* clonedNode = new Node(node->val);
        clonedNodes[node->val] = clonedNode;
        for(Node* neighbor : node->neighbors) {
            Node* clonedNeighbor = cloneDfs(neighbor, clonedNodes);
            clonedNode->neighbors.push_back(clonedNeighbor);
        }
        return clonedNode;
    }
};